### PR TITLE
[TensorExpr] Add a boilerplate pass for future TensorExpr fusion pass.

### PR DESCRIPTION
**Pull request by ZolotukhinM**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33464 [TensorExpr] Add a boilerplate pass for future TensorExpr fusion pass.**

I added a python-exposed knob to register this pass in custom passes pipeline. If the knob is not used, the pass is not registered and thus not run at all.

Differential Revision: [D19958217](https://our.internmc.facebook.com/intern/diff/D19958217)

### CODE REVIEW

File: torch/csrc/jit/passes/tensorexpr_fuser.h [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <memory>
+
+namespace torch {
+namespace jit {
+
+struct Graph;
+
+// Run TensorExpressions-based fuser.
+TORCH_API void fuseTensorExprs(std::shared_ptr<Graph>& graph);
```
Comment (bertmaher) [inline torch/csrc/jit/passes/tensorexpr_fuser.h]: Do we need to expose this in the header?  Not that it's a big deal.

**Review by bertmaher — APPROVED** [review]
lgtm!

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+std::pair<graph_node_list::iterator, bool> scanNode(
+    Node* consumer,
+    AliasDb& aliasDb) {
+  auto inputs =
+      sortReverseTopological(consumer->inputs(), consumer->owningBlock());
+
+  // Grab the iterator below consumer.  We'll use that to determine
+  // where to resume iteration, even if consumer gets relocated within
+  // the block.
+  auto iter = --consumer->reverseIterator();
```
Comment (suo) [inline torch/csrc/jit/passes/tensorexpr_fuser.cpp]: I find the iterator flow here a little difficult to understand—decrementing a reverse iterator, double incrementing an iterator, doing it prefix, etc. Is there some way we can use the `next()`/`prev()` APIs to make this a little clearer?

**Review by suo — APPROVED** [review]
lgtm as a skeleton. We may need to revise post-functionalization pass (cc @eellison) but that should not block this

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp (hunk 2 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/passes/tensorexpr_fuser.cpp]: There should be an easier way, yes - e.g. `create_autodiff_subgraph` does a similar thing and doesn't have this decrements/double-increments. I will follow up on this as we start adding actual code.

PS: I also hope this all becomes much simpler once we have functional subgraphs.

File: torch/csrc/jit/passes/tensorexpr_fuser.h (hunk 1 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/passes/tensorexpr_fuser.h]: I think it's good to have it exposed at least for testing purposes.

**Pull request closed without merging**